turbopack/crates/turbopack-ecmascript/src/references/esm/base.rs

```diff
@@ -51,18 +58,181 @@ pub enum ReferencedAsset {
+        scope_hoisting_context: Option<ScopeHoistingContext<'_>>,
+    ) -> Result<Option<ReferencedAssetIdent>> {
         Ok(match self {
             ReferencedAsset::Some(asset) => {
-                Some(Self::get_ident_from_placeable(asset, chunking_context).await?)
+                if let Some(scope_hoisting_context) = scope_hoisting_context {
+                    if let Some(ctxt) = scope_hoisting_context
+                        .module_syntax_contexts
+                        .get(&ResolvedVc::upcast(*asset))
+                    {
+                        if let Some(export) = &export {
+                            if let EcmascriptExports::EsmExports(exports) =
+                                *asset.get_exports().await?
+                            {
+                                let exports = exports.expand_exports(None).await?;
+                                let export = exports.exports.get(export);
+                                match export {
+                                    Some(EsmExport::LocalBinding(name, _)) => {
+                                        // A local binding in a module that is merged in the same
+                                        // group
+                                        return Ok(Some(ReferencedAssetIdent::LocalBinding {
+                                            ident: name.clone(),
+                                            ctxt: *ctxt,
+                                        }));
+                                    }
+                                    Some(b @ EsmExport::ImportedBinding(esm_ref, _, _))
+                                    | Some(b @ EsmExport::ImportedNamespace(esm_ref)) => {
+                                        let export =
+                                            if let EsmExport::ImportedBinding(_, export, _) = b {
+                                                Some(export.clone())
+                                            } else {
+                                                None
+                                            };
+
+                                        let referenced_asset =
+                                            ReferencedAsset::from_resolve_result(
+                                                esm_ref.resolve_reference(),
+                                            )
+                                            .await?;
+
+                                        // If the target module is still in the same group, we can
+                                        // refer it locally, otherwise it will be imported
+                                        return Ok(
+                                            match Box::pin(referenced_asset.get_ident(
+                                                chunking_context,
+                                                export,
+                                                Some(scope_hoisting_context),
+                                            ))
```

Comment:
This might cause infinite recursion on reexport cycles. see https://github.com/vercel/next.js/pull/79792